src/libpmemobj/pool.c

```diff
@@ -0,0 +1,195 @@
+ */
+struct pmalloc_pool *
+pool_new(void *ptr, size_t size, enum backend_type type)
+{
+	struct pmalloc_pool *pool = Malloc(sizeof (*pool));
+	if (pool == NULL) {
+		goto error_pool_malloc;
+	}
+
+	pool->backend = backend_open(type, ptr, size);
+	if (pool->backend == NULL) {
+		goto error_backend_new;
+	}
+
+	pool->lock = Malloc(sizeof (*pool->lock));
+	if (pool->lock == NULL) {
+		goto error_lock_malloc;
+	}
+
+	if (pthread_mutex_init(pool->lock, NULL) != 0) {
+		goto error_lock_init;
+	}
+
+	memset(pool->buckets, 0, sizeof (pool->buckets));
+	pool->p_ops = pool->backend->p_ops;
+
+	return pool;
+
+error_lock_init:
+	Free(pool->lock);
+error_lock_malloc:
+	backend_close(pool->backend);
+error_backend_new:
+	Free(pool);
+error_pool_malloc:
+	return NULL;
+}
+
+/*
+ * pool_delete -- deinitialize and free pool object
+ */
+void
+pool_delete(struct pmalloc_pool *p)
+{
+	for (int i = 0; i < MAX_BUCKETS; ++i) {
+		if (p->buckets[i] != NULL)
+			bucket_delete(p->buckets[i]);
+	}
+
+	for (int i = 0; i < MAX_ARENAS; ++i) {
+		if (p->arenas[i] != NULL) {
+			arena_delete(p->arenas[i]);
+		}
+	}
+
+	if (pthread_mutex_destroy(p->lock) != 0) {
+		LOG(4, "Failed to destroy pool lock");
+	}
+
+	backend_close(p->backend);
+
+	Free(p);
+}
+
+/*
+ * select_arena_id -- (internal) finds a least-used arena
+ */
+static int
+select_arena_id(struct pmalloc_pool *p)
+{
+	int min_arena_threads = ~0;
+	int id = -1;
+	for (int i = 0; i < MAX_ARENAS; ++i) {
+		if (p->arenas[i] == NULL) {
+			id = i;
+			break;
+		}
+
+		if (p->arenas[i]->associated_threads < min_arena_threads) {
+			id = i;
+			min_arena_threads = p->arenas[i]->associated_threads;
+		}
+	}
+
+	return id;
+}
+
+/*
+ * select_thread_arena_slow -- (internal) slow path of arena selection
+ */
+static struct arena *
+select_thread_arena_slow(struct pmalloc_pool *p)
+{
+	if (pthread_mutex_lock(p->lock) != 0) {
+		LOG(4, "Failed to acquire pool mutex");
+		return NULL;
+	}
+
+	if (arena_id == -1) {
+		arena_id = select_arena_id(p);
+	}
+
+	if (p->arenas[arena_id] == NULL) {
+		if ((p->arenas[arena_id] = arena_new(p, arena_id)) == NULL) {
+			goto error_arena_new;
+		}
+	}
+
+	p->arenas[arena_id]->associated_threads += 1;
+
+error_arena_new:
+	if (pthread_mutex_unlock(p->lock) != 0) {
+		LOG(4, "Failed to release pool mutex");
+	}
+
+	return p->arenas[arena_id];
+}
+
+/*
+ * pool_select_arena -- select an arena associated with current thread
+ */
+struct arena *
+pool_select_arena(struct pmalloc_pool *p)
+{
+	return (arena_id == -1 || p->arenas[arena_id] == NULL) ?
+		select_thread_arena_slow(p) :
+		p->arenas[arena_id];
+}
+
+/*
+ * pool_recycle_object -- adds the object to the global pool bucket
+ */
+bool
+pool_recycle_object(struct pmalloc_pool *p, struct bucket_object *obj)
+{
+	int class_id = get_bucket_class_id_by_size(p, obj->real_size);
+	if (p->buckets[class_id] == NULL) {
+		if ((p->buckets[class_id] = bucket_new(p, class_id)) == NULL)
+			return false;
+	}
```

Comment:
This is optimization for the 99% case in which a single application will only ever use a single pool. I've spent some time pondering about possible solutions that would allow for unlimited number of pools AND always used optimial arena for the pool, I came to the conclusion that the complexity of this solution would be greater than the benefits. Also, notice the fact that I do not use pthreads destructor functions do decrease the "associated_threads" number so arenas are effiectively choosen in a round robin fashion.